diff -Naur orig/lib/lv_bindings/lv_conf.h new/lib/lv_bindings/lv_conf.h
--- orig/lib/lv_bindings/lv_conf.h	2021-03-19 12:51:05.033997501 +0100
+++ new/lib/lv_bindings/lv_conf.h	2021-03-19 15:11:13.849714560 +0100
@@ -19,7 +19,7 @@
  *====================*/
 
 /* Maximal horizontal and vertical resolution to support by the library.*/
-#define LV_HOR_RES_MAX          (480)
+#define LV_HOR_RES_MAX          (240)
 #define LV_VER_RES_MAX          (320)
 
 /* Color depth:
@@ -29,13 +29,13 @@
  * - 32: ARGB8888
  */
 #ifndef LV_COLOR_DEPTH
-#define LV_COLOR_DEPTH     32
+#define LV_COLOR_DEPTH     16
 #endif
 
 /* Swap the 2 bytes of RGB565 color.
  * Useful if the display has a 8 bit interface (e.g. SPI)*/
 #ifndef LV_COLOR_16_SWAP
-#define LV_COLOR_16_SWAP   0
+#define LV_COLOR_16_SWAP   1
 #endif
 
 /* 1: Enable screen transparency.
@@ -406,7 +406,7 @@
 /* Demonstrate special features */
 #define LV_FONT_MONTSERRAT_12_SUBPX      0
 #define LV_FONT_MONTSERRAT_28_COMPRESSED 0  /*bpp = 3*/
-#define LV_FONT_DEJAVU_16_PERSIAN_HEBREW 1  /*Hebrew, Arabic, PErisan letters and all their forms*/
+#define LV_FONT_DEJAVU_16_PERSIAN_HEBREW 0  /*Hebrew, Arabic, PErisan letters and all their forms*/
 #define LV_FONT_SIMSUN_16_CJK            0  /*1000 most common CJK radicals*/
 
 /*Pixel perfect monospace font
@@ -453,11 +453,11 @@
 
 /* No theme, you can apply your styles as you need
  * No flags. Set LV_THEME_DEFAULT_FLAG 0 */
-#define LV_USE_THEME_EMPTY       1
+#define LV_USE_THEME_EMPTY       0
 
 /*Simple to the create your theme based on it
  * No flags. Set LV_THEME_DEFAULT_FLAG 0 */
-#define LV_USE_THEME_TEMPLATE    1
+#define LV_USE_THEME_TEMPLATE    0
 
 /* A fast and impressive theme.
  * Flags:
@@ -473,17 +473,17 @@
  * texts and borders will be black and the background will be
  * white. Else the colors are inverted.
  * No flags. Set LV_THEME_DEFAULT_FLAG 0 */
-#define LV_USE_THEME_MONO        1
+#define LV_USE_THEME_MONO        0
 
 #define LV_THEME_DEFAULT_INCLUDE            <stdint.h>      /*Include a header for the init. function*/
 #define LV_THEME_DEFAULT_INIT               lv_theme_material_init
-#define LV_THEME_DEFAULT_COLOR_PRIMARY      lv_color_hex(0x01a2b1)
-#define LV_THEME_DEFAULT_COLOR_SECONDARY    lv_color_hex(0x44d1b6)
-#define LV_THEME_DEFAULT_FLAG               LV_THEME_MATERIAL_FLAG_LIGHT
+#define LV_THEME_DEFAULT_COLOR_PRIMARY      lv_color_hex(0x0074d9)
+#define LV_THEME_DEFAULT_COLOR_SECONDARY    lv_color_hex(0x0360b1)
+#define LV_THEME_DEFAULT_FLAG               LV_THEME_MATERIAL_FLAG_LIGHT | LV_THEME_MATERIAL_FLAG_NO_FOCUS
 #define LV_THEME_DEFAULT_FONT_SMALL         &lv_font_montserrat_14
 #define LV_THEME_DEFAULT_FONT_NORMAL        &lv_font_montserrat_14
 #define LV_THEME_DEFAULT_FONT_SUBTITLE      &lv_font_montserrat_14
-#define LV_THEME_DEFAULT_FONT_TITLE         &lv_font_montserrat_14
+#define LV_THEME_DEFAULT_FONT_TITLE         &lv_font_montserrat_16
 
 /*=================
  *  Text settings
diff -Naur orig/lib/lv_bindings/lvgl/src/lv_themes/lv_theme_material.c new/lib/lv_bindings/lvgl/src/lv_themes/lv_theme_material.c
--- orig/lib/lv_bindings/lvgl/src/lv_themes/lv_theme_material.c	2021-03-19 12:50:47.545770462 +0100
+++ new/lib/lv_bindings/lvgl/src/lv_themes/lv_theme_material.c	2021-03-19 12:47:00.415009997 +0100
@@ -18,8 +18,8 @@
 
 /*SCREEN*/
 
-#define COLOR_SCR        (IS_LIGHT ? lv_color_hex(0xeaeff3) : lv_color_hex(0x444b5a))
-#define COLOR_SCR_TEXT   (IS_LIGHT ? lv_color_hex(0x3b3e42) : lv_color_hex(0xe7e9ec))
+#define COLOR_SCR        (IS_LIGHT ? lv_color_hex(0xffffff) : lv_color_hex(0x444b5a))
+#define COLOR_SCR_TEXT   (IS_LIGHT ? lv_color_hex(0x000000) : lv_color_hex(0xe7e9ec))
 
 /*BUTTON*/
 #define COLOR_BTN           (IS_LIGHT ? lv_color_hex(0xffffff) : lv_color_hex(0x586273))
@@ -257,7 +257,7 @@
     lv_style_set_transition_prop_5(&styles->bg_click, LV_STATE_DEFAULT, LV_STYLE_BG_COLOR);
 
     lv_style_reset(&styles->btn);
-    lv_style_set_radius(&styles->btn, LV_STATE_DEFAULT, LV_RADIUS_CIRCLE);
+    lv_style_set_radius(&styles->btn, LV_STATE_DEFAULT, LV_DPX(8)); //TH LV_RADIUS_CIRCLE);
     lv_style_set_bg_opa(&styles->btn, LV_STATE_DEFAULT, LV_OPA_COVER);
     lv_style_set_bg_color(&styles->btn, LV_STATE_DEFAULT, COLOR_BTN);
     lv_style_set_bg_color(&styles->btn, LV_STATE_PRESSED, COLOR_BTN_PR);
@@ -291,8 +291,8 @@
 
     lv_style_set_pad_left(&styles->btn, LV_STATE_DEFAULT, LV_DPX(40));
     lv_style_set_pad_right(&styles->btn, LV_STATE_DEFAULT, LV_DPX(40));
-    lv_style_set_pad_top(&styles->btn, LV_STATE_DEFAULT, LV_DPX(15));
-    lv_style_set_pad_bottom(&styles->btn, LV_STATE_DEFAULT, LV_DPX(15));
+    lv_style_set_pad_top(&styles->btn, LV_STATE_DEFAULT, LV_DPX(4));  //TH LV_DPX(15));
+    lv_style_set_pad_bottom(&styles->btn, LV_STATE_DEFAULT, LV_DPX(4)); //TH LV_DPX(15));
     lv_style_set_pad_inner(&styles->btn, LV_STATE_DEFAULT, LV_DPX(20));
     lv_style_set_outline_width(&styles->btn, LV_STATE_DEFAULT, OUTLINE_WIDTH);
     lv_style_set_outline_opa(&styles->btn, LV_STATE_DEFAULT, LV_OPA_0);
@@ -836,26 +836,29 @@
 #if LV_USE_TABVIEW || LV_USE_WIN
     lv_style_reset(&styles->tabview_btns_bg);
     lv_style_set_bg_opa(&styles->tabview_btns_bg, LV_STATE_DEFAULT, LV_OPA_COVER);
-    lv_style_set_bg_color(&styles->tabview_btns_bg, LV_STATE_DEFAULT, COLOR_BG);
-    lv_style_set_text_color(&styles->tabview_btns_bg, LV_STATE_DEFAULT, COLOR_SCR_TEXT);
-    lv_style_set_image_recolor(&styles->tabview_btns_bg, LV_STATE_DEFAULT, lv_color_hex(0x979a9f));
+    lv_style_set_bg_color(&styles->tabview_btns_bg, LV_STATE_DEFAULT, theme.color_primary);   //TH
+    lv_style_set_text_color(&styles->tabview_btns_bg, LV_STATE_DEFAULT, COLOR_BG);  //TH
+    lv_style_set_image_recolor(&styles->tabview_btns_bg, LV_STATE_DEFAULT, lv_color_hex(0xffffff));
     lv_style_set_pad_top(&styles->tabview_btns_bg, LV_STATE_DEFAULT, LV_DPX(7));
     lv_style_set_pad_left(&styles->tabview_btns_bg, LV_STATE_DEFAULT, LV_DPX(7));
     lv_style_set_pad_right(&styles->tabview_btns_bg, LV_STATE_DEFAULT, LV_DPX(7));
-
+    lv_style_set_text_font(&styles->tabview_btns_bg, LV_STATE_DEFAULT, theme.font_title); //TH
+    
     lv_style_reset(&styles->tabview_btns);
     lv_style_set_bg_opa(&styles->tabview_btns, LV_STATE_PRESSED, LV_OPA_50);
-    lv_style_set_bg_color(&styles->tabview_btns, LV_STATE_PRESSED, lv_color_hex3(0x888));
-    lv_style_set_text_color(&styles->tabview_btns, LV_STATE_CHECKED, COLOR_SCR_TEXT);
-    lv_style_set_pad_top(&styles->tabview_btns, LV_STATE_DEFAULT, LV_DPX(20));
+    lv_style_set_bg_color(&styles->tabview_btns, LV_STATE_PRESSED, theme.color_secondary);  //TH
+    lv_style_set_text_color(&styles->tabview_btns, LV_STATE_DEFAULT, lv_color_hex(0xdddddd)); //TH
+    lv_style_set_image_recolor(&styles->tabview_btns, LV_STATE_DISABLED, lv_color_hex(0xa0a0a0)); //TH
+    lv_style_set_text_color(&styles->tabview_btns, LV_STATE_CHECKED, COLOR_BG); //TH SCR_TEXT);
+    lv_style_set_pad_top(&styles->tabview_btns, LV_STATE_DEFAULT, LV_DPX(20));  //TH
     lv_style_set_pad_bottom(&styles->tabview_btns, LV_STATE_DEFAULT, LV_DPX(20));
     lv_style_set_text_color(&styles->tabview_btns, LV_STATE_FOCUSED, theme.color_primary);
     lv_style_set_text_color(&styles->tabview_btns, LV_STATE_EDITED, theme.color_secondary);
 
     lv_style_reset(&styles->tabview_indic);
     lv_style_set_bg_opa(&styles->tabview_indic, LV_STATE_DEFAULT, LV_OPA_COVER);
-    lv_style_set_bg_color(&styles->tabview_indic, LV_STATE_DEFAULT, theme.color_primary);
-    lv_style_set_bg_color(&styles->tabview_indic, LV_STATE_EDITED, theme.color_secondary);
+    lv_style_set_bg_color(&styles->tabview_indic, LV_STATE_DEFAULT, lv_color_hex3(0xaaa)); //TH
+    lv_style_set_bg_color(&styles->tabview_indic, LV_STATE_EDITED, theme.color_primary); //TH
     lv_style_set_size(&styles->tabview_indic, LV_STATE_DEFAULT, LV_DPX(5));
     lv_style_set_radius(&styles->tabview_indic, LV_STATE_DEFAULT, LV_RADIUS_CIRCLE);
 
@@ -1148,7 +1151,7 @@
 #if LV_USE_TABVIEW
         case LV_THEME_TABVIEW:
             list = lv_obj_get_style_list(obj, LV_TABVIEW_PART_BG);
-            _lv_style_list_add_style(list, &styles->scr);
+	    _lv_style_list_add_style(list, &styles->scr);
 
             list = lv_obj_get_style_list(obj, LV_TABVIEW_PART_TAB_BG);
             _lv_style_list_add_style(list, &styles->tabview_btns_bg);
diff -Naur orig/ports/esp32/boards/sdkconfig.spiram new/ports/esp32/boards/sdkconfig.spiram
--- orig/ports/esp32/boards/sdkconfig.spiram	2021-03-19 12:56:08.478100648 +0100
+++ new/ports/esp32/boards/sdkconfig.spiram	2021-03-19 12:55:53.153888455 +0100
@@ -7,3 +7,13 @@
 
 # v3.3-only (renamed in 4.0)
 CONFIG_SPIRAM_SUPPORT=y
+
+CONFIG_ESPTOOLPY_FLASHMODE_QIO=y
+CONFIG_ESPTOOLPY_FLASHMODE="qio"
+
+CONFIG_ESPTOOLPY_FLASHFREQ_80M=y
+CONFIG_ESPTOOLPY_FLASHFREQ="80m"
+
+CONFIG_FLASHMODE_QIO=y
+CONFIG_SPIRAM_SPEED_80M=y
+CONFIG_ESP32_REV_MIN_1=y
